### PR TITLE
Add epicli versioning (#1306)

### PR DESCRIPTION
**Pull request by sbbroot**
* Add version info file creation for each VM used.

### CODE REVIEW

**Comment by seriva** [comment]
First we need to merge #2643.

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+- name: Create versioning directory if it does not exist.
```
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]:
```suggestion
- name: Create versioning directory if it does not exist
```

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  file:
+    path: /etc/epicli
+    state: directory
+    mode: '0755'
```
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]:
```suggestion
    mode: u=rwx,go=rx
```

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+- name: Get versioning file status.
```
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]:
```suggestion
- name: Get versioning file status
```

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  stat:
+    path: /etc/epicli/version
+  register: versioning_file
+
+- name: Create versioning file.
```
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]:
```suggestion
- name: Create versioning file
```

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  copy:
+    dest: /etc/epicli/version
+    content: |
+      #mode | date | version
+  when: not versioning_file.stat.exists
+
+- name: Append versioning info.
```
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]:
```suggestion
- name: Append versioning info
```

File: core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml (hunk 2 above)
Comment (atsikham) [inline core/src/epicli/data/common/ansible/playbooks/roles/versioning/tasks/main.yml]: Has it been checked by linter?

File: core/src/epicli/cli/engine/ansible/AnsibleRunner.py [hunk 6]
```diff
@@ -152,6 +161,12 @@ def upgrade(self):
         # pre-flight to prepare machines
         self.pre_flight(inventory_path)
 
+        self.ansible_command.run_playbook(inventory=inventory_path,
+                                          playbook_path=self.playbook_path(to_role_name('versioning')),
+                                          extra_vars={'mode': 'upgrade',
+                                                      'date': datetime.datetime.now().ctime(),
+                                                      'version': VERSION})
```
Comment (atsikham) [inline core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: There is `is_upgrade_run` fact and `ansible_date_time` can be used from ansible. Is it necessary to pass these parameters and modify `run_playbook ` for that?

**Review by atsikham — CHANGES_REQUESTED** [review]

File: core/src/epicli/cli/engine/ansible/AnsibleRunner.py (hunk 6 above)
Comment (to-bar) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: Yes, there can be also time differences, remote time (on VM) `ansible_date_time` can be different than local time (on Ansible controller).

**Review by to-bar — CHANGES_REQUESTED** [review]
IMO:
1) We may no need separate playbook. `postflight` playbook can be used if we want to log only successful runs.
2) We can use parseable file format, for example YAML (Ansible has `from_yaml` and `to_nice_yaml` filters).
Sample structure:
```
# ordered list of successful epicli runs - the latest at the top
deployments:
  - version: 1.2.0
    date: 2021-08-17 15:24
    mode: apply
    state: started
  - version: 1.2.0
    date: 2021-08-17 15:24
    mode: upgrade
    state: completed
  - version: 1.1.0
    date: 2021-05-17 15:24
    mode: apply
    state: completed
```
3) `/etc/epicli/version` should be replaced with something like `/var/lib/epiphany/history.yml`. `\etc` is mainly intended for configuration files, see [here](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html).
4) If we implement deployment history instead of just the latest version, then role can be renamed to e.g. `deployment_history`.

**Comment by atsikham** [comment]
What about adding the status for each run? We can add info in preflight about started run and update state in postflight. This way if there is a failure and partial installation is done, we will have an information about that.

File: core/src/epicli/cli/engine/ansible/AnsibleRunner.py (hunk 6 above)
Comment (sbbroot) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: I do not agree with the `is_upgrade_run` change. If we want to add something later we will probably end up with a bunch of `if` `else`s. Here the script doesn't care about which mode it is, so there is no place for conditionals.

Comment (sbbroot) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: About `ansible_date_time`, maybe we should add both local and remote time?

Comment (atsikham) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: We don't need to implement `is_upgrade_run`, it's already in place and used in Ansible configuration.

Comment (sbbroot) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: I meant that it's a bad design, and if there is a better way of doing it, we should do it. Scripts should not be aware of the mode they are in, this leads to putting conditionals everywhere. In our python code we have distinct paths for both apply and upgrade modes.

Comment (atsikham) [reply core/src/epicli/cli/engine/ansible/AnsibleRunner.py]: Ok, before talking about design, let’s change the code not to use a separate role as @to-bar suggested. Then we will see what’s necessary in addition.

File: cli/engine/ansible/AnsibleRunner.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
+import datetime
 import os
 import time
+import shutil
```
Comment (atsikham) [inline cli/engine/ansible/AnsibleRunner.py]: Is this change necessary without changes in code?

File: ansible/playbooks/postflight.yml [hunk 8]
```diff
@@ -5,3 +5,18 @@
   become_method: sudo
   roles:
     - postflight
+
+
```
Comment (atsikham) [inline ansible/playbooks/postflight.yml]: Remove one empty line.

File: ansible/playbooks/postflight.yml [hunk 9]
```diff
@@ -5,3 +5,18 @@
...
+- name: Ansible playbook for changing the state of the current running version info
```
Comment (atsikham) [inline ansible/playbooks/postflight.yml]: - It's not necessary to mention that this is Ansible playbook.
- `current running version` - looks not good? Whoose version?

May be something like, just a suggestion: `Change the state of the previous Epicli execution`

File: ansible/playbooks/postflight.yml [hunk 10]
```diff
@@ -5,3 +5,18 @@
...
+  hosts: all
+  gather_facts: yes
```
Comment (atsikham) [inline ansible/playbooks/postflight.yml]:
```suggestion
  gather_facts: true
```

File: ansible/playbooks/postflight.yml [hunk 11]
```diff
@@ -5,3 +5,18 @@
...
+  become: true
+  become_method: sudo
+  tasks:
+    - name: Change current running status to complete
+      lineinfile:
+        path: /var/lib/epiphany/history.yml
+        firstmatch: yes  # only first occurrence
```
Comment (atsikham) [inline ansible/playbooks/postflight.yml]:
```suggestion
        firstmatch: true  # only first occurrence
```

File: ansible/playbooks/postflight.yml [hunk 12]
```diff
@@ -5,3 +5,18 @@
...
+        backrefs: yes  # allows to use regexp groups
```
Comment (atsikham) [inline ansible/playbooks/postflight.yml]:
```suggestion
        backrefs: true  # allows to use regexp groups
```

File: ansible/playbooks/postflight.yml (hunk 11 above)
Comment (atsikham) [inline ansible/playbooks/postflight.yml]: May be there a situation when such update will change not the last status? Is there a way tostart  search from the end?

File: ansible/playbooks/preflight.yml [hunk 13]
```diff
@@ -11,3 +11,50 @@
     - preflight
   environment:
     KUBECONFIG: "{{ kubeconfig.remote }}"
+
+
+- name: Ansible playbook for adding epicli version info for each component used
```
Comment (atsikham) [inline ansible/playbooks/preflight.yml]: Same as previous.

File: ansible/playbooks/preflight.yml [hunk 14]
```diff
@@ -11,3 +11,50 @@
...
+  hosts: all
+  gather_facts: yes
```
Comment (atsikham) [inline ansible/playbooks/preflight.yml]: Change all `yes` values to `true`

File: cli/engine/ansible/AnsibleVarsGenerator.py [hunk 15]
```diff
@@ -142,6 +142,7 @@ def populate_group_vars(self, ansible_dir):
         main_vars['is_upgrade_run'] = self.is_upgrade_run
         main_vars['roles_with_generated_vars'] = sorted(self.roles_with_generated_vars)
         main_vars['upgrade_components'] = Config().upgrade_components
+        main_vars['version'] = VERSION
```
Comment (atsikham) [inline cli/engine/ansible/AnsibleVarsGenerator.py]: Looks like not required thing - there is a version in generated defaults. @to-bar what do you think?

**Review by atsikham — DISMISSED** [review]

File: ansible/playbooks/postflight.yml (hunk 11 above)
Comment (sbbroot) [reply ansible/playbooks/postflight.yml]: I don't think that there is a possibility that such thing could happen.
Preflight task always puts newest execution on top.

File: cli/engine/ansible/AnsibleVarsGenerator.py (hunk 15 above)
Comment (sbbroot) [reply cli/engine/ansible/AnsibleVarsGenerator.py]: Version value is not visible at preflight stage without this change.

Comment (to-bar) [reply cli/engine/ansible/AnsibleVarsGenerator.py]: For me this is ok since VERSION is the single source of true.  But I would change `main_vars['version']` to `main_vars['epiphany_version']` for two reasons: 1) more intuitive, 2) lower risk that it will be overridden for example by `set_fact` task.

File: ansible/playbooks/postflight.yml [hunk 16]
```diff
@@ -5,3 +5,17 @@
   become_method: sudo
   roles:
     - postflight
+
+- name: Change the state info of the previous epicli execution
+  hosts: all
+  gather_facts: true
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
  gather_facts: false
```
It looks like facts are not used here but gathering them takes time.

File: ansible/playbooks/postflight.yml [hunk 17]
```diff
@@ -5,3 +5,17 @@
...
+  become: true
+  become_method: sudo
+  tasks:
+    - name: Change current running status to complete
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]: You can consider more robust (but longer) approach:
1) Load file using `slurp` module
2) Parse yaml using Ansible filter (`from_yaml`)
3) Update value using Ansible filter(s) - probably `combine`
4) Save updated contents to file (using `copy` module)
Example: https://github.com/epiphany-platform/epiphany/blob/7e7987ab6d3dfd047040c487b4d4a1c230887ba4/ansible/playbooks/roles/kubernetes_common/tasks/extend-kubeadm-config.yml#L19

File: ansible/playbooks/preflight.yml [hunk 18]
```diff
@@ -11,3 +11,50 @@
...
+      lineinfile:
+        path: /var/lib/epiphany/history.yml
+        firstmatch: true  # only first occurrence
+        backrefs: true  # allows to use regexp groups
+        regexp: '(.*)state\: started'  # assuming that current status is always on top
+        line: '\1state: failed'
+      when: history_file.stat.exists and history_file.stat.size > 0
+
+    - name: Add history info
+      lineinfile:
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]: Operations on lists - like prepend are easier to understand than using regexp that replaces single line string with multi-line.

File: ansible/playbooks/preflight.yml [hunk 19]
```diff
@@ -11,3 +11,50 @@
...
+        path: /var/lib/epiphany/history.yml
+        regexp: 'deployments:'
+        line: "{{ history_file_content }}"
+      when: history_file.stat.exists and history_file.stat.size > 0
+
+    - name: Create history file
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]: Without `lineinfile` you would have full content stored in variable (`set_fact`) and the when condition will be unneeded.

**Review by to-bar — CHANGES_REQUESTED** [review]

File: ansible/playbooks/postflight.yml (hunk 16 above)
Comment (to-bar) [reply ansible/playbooks/postflight.yml]: This would make sense if we update not only the `state` but the `date` as well.

File: ansible/playbooks/postflight.yml (hunk 17 above)
Comment (sbbroot) [reply ansible/playbooks/postflight.yml]: I find it overengineered. I don't see robustness through putting more code (and possibly more bugs).

File: ansible/playbooks/preflight.yml (hunk 18 above)
Comment (sbbroot) [reply ansible/playbooks/preflight.yml]: Current solution uses only single variable for each task. Changing it would make it more complex in my opinion.

File: ansible/playbooks/postflight.yml (hunk 17 above)
Comment (to-bar) [reply ansible/playbooks/postflight.yml]: My opinion is that files that have a standardized format (like yml, json, xml) should be modified using parsers. And Ansible has built-in utils to work with YAML. So I will not approve regexp based solution.

**Review by erzetpe — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

File: ansible/playbooks/preflight.yml [hunk 20]
```diff
@@ -11,3 +11,57 @@
     - preflight
   environment:
     KUBECONFIG: "{{ kubeconfig.remote }}"
+
+- name: Add epicli version info for each component used
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
- name: Add entry to epicli execution history on each host
```

File: ansible/playbooks/preflight.yml [hunk 21]
```diff
@@ -11,3 +11,57 @@
...
+  become: true
+  become_method: sudo
+  vars:
+    history_last_entry: {}
+    history_other_entries: []
+    history_newest_entry:
+        version: "{{ epiphany_version }}"
+        date: "{{ ansible_date_time.date }} {{ ansible_date_time.time }}"
+        mode: "{{ is_upgrade_run | ternary('upgrade', 'apply') }}"
+        state: started
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
      version: "{{ epiphany_version }}"
      date: "{{ ansible_date_time.date }} {{ ansible_date_time.time }}"
      mode: "{{ is_upgrade_run | ternary('upgrade', 'apply') }}"
      status: started
```

File: ansible/playbooks/preflight.yml [hunk 22]
```diff
@@ -11,3 +11,57 @@
...
+  tasks:
+    - name: Create epiphany directory if it does not exist
+      file:
+        path: /var/lib/epiphany/
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
        path: /var/lib/epiphany
```

File: ansible/playbooks/preflight.yml [hunk 23]
```diff
@@ -11,3 +11,57 @@
...
+      register: history_file_handler
+      when: history_file.stat.exists and history_file.stat.size > 0
+
+    - name: Parse history_file_handler content to yaml format
+      set_fact:
+        history_last_entry: "{{ history_file_yaml.deployments[0] }}"
+        history_other_entries: "{{ history_file_yaml.deployments[1:] }}"
+      vars:
+        history_file_yaml: "{{ history_file_handler.content | b64decode | from_yaml }}"
+      when: history_file.stat.exists and history_file.stat.size > 0
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
    - name: Parse history file
      when:
        - history_file.stat.exists
        - history_file.stat.size > 0
      block:
        - name: Load history file
          slurp:
            src: /var/lib/epiphany/history.yml
          register: history_file_handler

        - name: Parse history file
          set_fact:
            history_latest_entry: "{{ _history_file_yaml.deployments[0] }}"
            history_other_entries: "{{ _history_file_yaml.deployments[1:] }}"
          vars:
            _history_file_yaml: "{{ history_file_handler.content | b64decode | from_yaml }}"
```

File: ansible/playbooks/preflight.yml [hunk 24]
```diff
@@ -11,3 +11,57 @@
...
+
+    - name: Construct new history file
+      set_fact:
+        new_history_file: >-
+          {{ {'deployments': [history_newest_entry] + history_last_entry | ternary([history_last_entry], []) + history_other_entries} }}
+
+    - name: Save the history file content
+      copy:
+        content: "{{ new_history_file | to_nice_yaml }}"
+        dest: /var/lib/epiphany/history.yml
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
    - name: Check previous status and update if needed
      set_fact:
        history_latest_entry: "{{ history_latest_entry | combine({'status': 'failed'}) }}"
      when:
        - history_latest_entry.status is defined
        - history_latest_entry.status != 'completed'

    - name: Update history and save to file
      copy:
        content: "{{ _new_content | to_nice_yaml(indent=2) }}"
        dest: /var/lib/epiphany/history.yml
      vars:
        _new_content:
          deployments: >-
            {{ [history_new_entry] + history_latest_entry | ternary([history_latest_entry], []) + history_other_entries }}
```

File: ansible/playbooks/postflight.yml [hunk 25]
```diff
@@ -5,3 +5,33 @@
   become_method: sudo
   roles:
     - postflight
+
+- name: Change the state info of the previous epicli execution
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
- name: Update status in epicli execution history file
```

File: ansible/playbooks/postflight.yml [hunk 26]
```diff
@@ -5,3 +5,33 @@
...
+  gather_facts: false
+  become: true
+  become_method: sudo
+  tasks:
+    - name: Load the history file
+      slurp:
+        src: /var/lib/epiphany/history.yml
+      register: history_file
+
+    - name: Parse history_file to yaml format and change the state
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
    - name: Parse history file and update the status
```

File: ansible/playbooks/postflight.yml [hunk 27]
```diff
@@ -5,3 +5,33 @@
...
+
+    - name: Construct history file
+      set_fact:
+        new_history_file: >-
+          {{ {'deployments': [history_first_entry] + history_other_entries} }}
+
+    - name: Save newly constructed history file
+      copy:
+        content: "{{ new_history_file | to_nice_yaml }}"
+        dest: /var/lib/epiphany/history.yml
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
    - name: Save updated history to file
      copy:
        content: "{{ _new_content | to_nice_yaml(indent=2) }}"
        dest: /var/lib/epiphany/history.yml
      vars:
        _new_content:
          deployments: "{{ [history_latest_entry] + history_other_entries }}"
```

File: ansible/playbooks/postflight.yml [hunk 28]
```diff
@@ -5,3 +5,33 @@
...
+  become_method: sudo
+  tasks:
+    - name: Load the history file
+      slurp:
+        src: /var/lib/epiphany/history.yml
+      register: history_file
+
+    - name: Parse history_file to yaml format and change the state
+      set_fact:
+        history_first_entry: >-
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
        history_latest_entry: >-
```

File: ansible/playbooks/preflight.yml [hunk 29]
```diff
@@ -11,3 +11,57 @@
...
   environment:
     KUBECONFIG: "{{ kubeconfig.remote }}"
+
+- name: Add epicli version info for each component used
+  hosts: all
+  gather_facts: true
+  become: true
+  become_method: sudo
+  vars:
+    history_last_entry: {}
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
    history_latest_entry: {}
```

File: ansible/playbooks/preflight.yml [hunk 30]
```diff
@@ -11,3 +11,57 @@
...
+    history_other_entries: []
+    history_newest_entry:
```
Comment (to-bar) [inline ansible/playbooks/preflight.yml]:
```suggestion
    history_new_entry:
```

File: ansible/playbooks/postflight.yml [hunk 31]
```diff
@@ -5,3 +5,33 @@
...
+          {{ history_file_yaml.deployments[0] | combine({'state': 'completed'})}}
```
Comment (to-bar) [inline ansible/playbooks/postflight.yml]:
```suggestion
          {{ history_file_yaml.deployments[0] | combine({'status': 'completed'})}}
```

**Review by to-bar — CHANGES_REQUESTED** [review]

**Review by to-bar — APPROVED** [review]

**Review by erzetpe — APPROVED** [review]

**Review by przemyslavic — APPROVED** [review]
